File: ruby/tuid.rb
require 'securerandom'

class TUID
  attr_accessor :clock
  attr_accessor :last
  attr_accessor :seq
  attr_reader :nid

  private(:last, :seq, :nid, :clock)

  def initialize(nid: 0, clock: ->(){(Time.now.utc.to_f*1000000).to_i})
    @clock = clock
    @nid = nid & 0xff
    @seq = 0
    @last = 0
  end

  def call
    rb = SecureRandom.random_bytes(6).bytes
    us = @clock.()

    if us > @last
      @last = us
      @seq = 0
    else
      @seq += 1
      if @seq > 0xff
        @seq = 0
        @last += 1
      end
    end

    TUID.pack(rb, @last, @nid, @seq)
  end

  def self.pack(rb, t, nid, seq)
    tb = [t].pack('Q>').bytes
    tuid = []
    tuid << tb[0] << tb[1] << tb[2] << tb[3] << tb[4] << tb[5]
    packed = [0x4000800000 | tb[6] << 28 | ((tb[7] & 0xf0) << 20) | ((tb[7] & 0xf) << 18) | (seq << 10) | nid << 2 | (rb[0] & 0x3)].pack('Q>').bytes
    tuid << packed[3] << packed[4] << packed[5] << packed[6] << packed[7] << rb[1] << rb[2] << rb[3] << rb[4] << rb[5]
    format('%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x', *tuid)
  end

  class << self
    def call
      @tuid = @tuid || TUID.new(nid:255)
      @tuid.()
    end
  end

end

File: c#/Tuid.cs
using System;

namespace Tuid
{
    public sealed class Generator
    {
        private readonly object _lock = new object();
        private ulong _last = 0;
        private ulong _seq = 0;
        private readonly ulong _nid = 0;
        private readonly Func<ulong> _clock;
        private readonly Func<ulong> _random;

        public Generator(ulong nid = 255, Func<ulong> clock = null, Func<ulong> random = null)
        {
            _clock = clock;
            _random = random;

            if (clock == null)
                _clock = DefaultClock;
            if (random == null)
                _random = DefaultRandom;
            _nid = nid;
            if (_nid > 255) {
              _nid = _random() & 0xff;
            }
        }

        private static readonly DateTime Epoch =
            new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc).ToUniversalTime();

        private static ulong DefaultClock()
        {
            return (ulong) (DateTime.UtcNow.Subtract(Epoch).TotalMilliseconds * 1000F);
        }

        private static readonly Random Random = new Random();

        private static ulong DefaultRandom()
        {
            return
                ((ulong) Random.Next() << 48) ^
                ((ulong) Random.Next() << 32) ^
                ((ulong) Random.Next() << 16) ^
                ((ulong) Random.Next());
        }

        public string Generate()
        {
            var us = _clock();
            var rand1 = _random();
            lock (_lock)
            {
                if (us <= _last)
                {
                    if (_seq >= 0xff)
                    {
                        ++_last;
                        _seq = 0;
                    }
                    else
                    {
                        ++_seq;
                    }
                }
                else
                {
                    _seq = (rand1 >> 12) & 0xff;
                    _last = us;
                }
            
                var a = _last >> 32;
                var b = (_last >> 16) & 0xffff;
                var c = 0x4000 | ((_last >> 4) & 0x0fff);
                var d = 0x8000 | ((_last & 0xf) << 10) | (_seq << 2) | (_nodeId >> 6);
                var e = ((_nid & 0x3f) << 10) | (rand1 & 0x3ff);
                var f = rand1>>32;
                return $"{a:x08}-{b:x04}-{c:x04}-{d:x04}-{e:x04}{f:x08}";
            }
        }
    }
}
